net/dnscache/dnscache.go

```diff
@@ -523,6 +523,21 @@ func (dc *dialCall) raceDial(ctx context.Context, ips []netip.Addr) (net.Conn, e
 		return nil, errors.New("no IPs")
 	}
 
+	// Partition candidate list and then merge such that an IPv6 address is
+	// in the first spot if present, and then addresses are interleaved.
+	// This ensures that we're trying an IPv6 address first, then
+	// alternating between v4 and v6 in case one of the two networks is
+	// broken.
+	var iv4, iv6 []netip.Addr
```

Comment:
Won't a sort be better?

Reply:
Oh I see, you're alternating.

Reply:
I don't actually want these IPs to be sorted, since otherwise we'll direct the majority of the traffic to the lowest-sorting IP. I'd actually rather randomly shuffle these to spread the load around, though I figured that could be a follow-up.

Reply:
At least for now, keeping the original presumably-at-least-sorta-random order and then merging them such that there's an IPv6 addr first means that we're not making things any worse.

Reply:
I meant StableSort with a func that uses the protocol as the Less func.

Reply:
Yeah, StableSort SGTM. That also means we don't allocate two more slices. (Assuming we own ips here... don't remember)

Reply:
the issue with StableSort is that it'll make it so that all IPv6 addresses are tried first. We really do want to alternate between IPv6 and IPv4 I think.

We should also really simplify this logic, it's very hard to understand. Not that I know how to simplify just yet.